base/path.jl

```diff
@@ -248,16 +248,19 @@ function splitpath(p::String)
     return out
 end
 
-joinpath(path::AbstractString)::String = path
-
 if Sys.iswindows()
 
-function joinpath(path::AbstractString, paths::AbstractString...)::String
-    result_drive, result_path = splitdrive(path)
+function joinpath(paths::Union{Tuple, AbstractVector})::String
```

Comment:
Why not duck type this and accept any iterator instead of only `tuple` and `AbstractVector`?

Reply:
The single-argument `joinpath` is ambiguous between a single string-like argument which should be converted to a string and a single argument that iterates path components which should be iterated, converted to strings and joined with path separators. Suppose someone tries to call this on an exotic type that's meant to represent a string-like type that doesn't subtype `AbstractString`, then what? If this API was duck-typed, then it would try to treat the argument like a collection and iterate it and then join the results together as strings. That might work fine if the thing happens to iterate path components, but what if it iterates something else like individual characters or grapheme clusters (which are represented as strings)? It's much better to make this API clearly constrained without ambiguity: we know what to do if the argument is an array or a tuple. If people really want to call `joinpath` on other collections, we can add those in the future on an as-needed basis, but it just doesn't seem likely to be necessary.